edgehog-device-runtime-store/src/models/containers/container.rs

```diff
@@ -0,0 +1,488 @@
+// This file is part of Edgehog.
+//
+// Copyright 2024 - 2025 SECO Mind Srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Container models.
+
+use std::{fmt::Display, ops::Deref};
+
+use diesel::{
+    backend::Backend,
+    deserialize::{FromSql, FromSqlRow},
+    dsl::{exists, BareSelect, Eq, Filter},
+    expression::AsExpression,
+    select,
+    serialize::{IsNull, ToSql},
+    sql_types::Integer,
+    sqlite::Sqlite,
+    Associations, ExpressionMethods, Insertable, QueryDsl, Queryable, Selectable,
+};
+
+use crate::{
+    conversions::SqlUuid,
+    models::containers::{image::Image, network::Network, volume::Volume},
+    schema::containers::{
+        container_missing_images, container_missing_networks, container_missing_volumes, containers,
+    },
+};
+
+/// Container configuration.
+#[derive(Debug, Clone, Insertable, Queryable, Associations, Selectable, PartialEq, Eq)]
+#[diesel(table_name = crate::schema::containers::containers)]
+#[diesel(belongs_to(Image))]
+#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
+pub struct Container {
+    /// Unique id received from Edgehog.
+    pub id: SqlUuid,
+    /// Container id returned by the container engine.
+    pub local_id: Option<String>,
+    /// Unique id received from Edgehog.
+    pub image_id: Option<SqlUuid>,
+    /// Status of the volume.
+    pub status: ContainerStatus,
+    /// Container network mode: none, bridge, ...
+    pub network_mode: String,
+    /// Hostname for the container
+    pub hostname: String,
+    /// Container restart policy
+    pub restart_policy: ContainerRestartPolicy,
+    /// Privileged
+    pub privileged: bool,
+}
+
+type ContainerById<'a> = Eq<containers::id, &'a SqlUuid>;
+type ContainerFilterById<'a> = Filter<containers::table, ContainerById<'a>>;
+type ContainerExists<'a> = BareSelect<exists<Filter<containers::table, ContainerById<'a>>>>;
+
+impl Container {
+    /// Returns the filter container table by id.
+    pub fn by_id(id: &SqlUuid) -> ContainerById<'_> {
+        containers::id.eq(id)
+    }
+
+    /// Returns the filtered container table by id.
+    pub fn find_id(id: &SqlUuid) -> ContainerFilterById<'_> {
+        containers::table.filter(Self::by_id(id))
+    }
+
+    /// Returns the container exists query.
+    pub fn exists(id: &SqlUuid) -> ContainerExists<'_> {
+        select(exists(Self::find_id(id)))
+    }
```

Comment:
Since each table of db is represented by a struct with these methods, What do you think to create a trait with these methods, and for the result of each methods a specific associated types?
Such as:
```rust
trait DbQuery {
 type A;
 type B;
 type C;

  pub fn by_id(id: &SqlUuid) -> A;
  pub fn find_id(id: &SqlUuid) -> B ;
  pub fn exists(id: &SqlUuid) -> C;
	
```
It seems something that you can generalize...

Reply:
I think it's possible. I will try it 👍🏻

Reply:
Done, except for the `exists` I created a default implementation. I cannot come up with the correct trait bound for it.